File: apps/api/app/schemas/document.py
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.schemas.audit import AuditLogRead


class OCRJobRead(BaseModel):
    id: str
    document_id: str
    job_type: str = "ocr"
    status: str
    attempts: int
    max_attempts: int = 3
    reason: str | None = None
    failed_reason: str | None = None
    error_message: str | None = None
    next_run_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DocumentRead(BaseModel):
    id: str
    title: str
    original_filename: str
    content_type: str | None = None
    document_type: str
    classification_confidence: float | None = None
    document_number: str | None = None
    document_symbol: str | None = None
    issued_date: date | None = None
    issued_place: str | None = None
    issuing_agency: str | None = None
    excerpt: str | None = None
    recipient: str | None = None
    signer_name: str | None = None
    signer_title: str | None = None
    seals_present: bool | None = None
    attachment_present: bool | None = None
    page_count: int | None = None
    metadata_source: str | None = None
    metadata_reviewed_at: datetime | None = None
    business_type: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DocumentListItemRead(DocumentRead):
    missing_module_metadata: bool = False


class DocumentPageRead(BaseModel):
    id: str
    page_number: int
    text: str
    confidence: float
    status: str

    model_config = {"from_attributes": True}


class DocumentFileRead(BaseModel):
    id: str
    document_id: str
    original_filename: str
    content_type: str | None = None
    file_size: int
    file_order: int
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DocumentChunkRead(BaseModel):
    id: str
    chunk_index: int
    text: str
    page_from: int | None = None
    page_to: int | None = None
    section_title: str | None = None
    doc_group: str | None = None
    chunk_level: str | None = None
    section_role: str | None = None
    section_path: list[str] | None = None
    chunk_confidence: float | None = None
    requires_review: bool = False
    qdrant_point_id: str | None = None

    model_config = {"from_attributes": True}


class ReviewQueueChunkRead(BaseModel):
    id: str
    document_id: str
    document_title: str
    document_number: str | None = None
    issued_date: date | None = None
    business_type: str | None = None
    chunk_index: int
    text: str
    page_from: int | None = None
    page_to: int | None = None
    section_title: str | None = None
    doc_group: str | None = None
    chunk_level: str | None = None
    section_role: str | None = None
    section_path: list[str] | None = None
    chunk_confidence: float | None = None
    requires_review: bool = True
    created_at: datetime
    updated_at: datetime


class ReviewQueueResponse(BaseModel):
    items: list[ReviewQueueChunkRead]
    total: int
    limit: int
    offset: int


class DocumentListResponse(BaseModel):
    items: list[DocumentListItemRead]
    total: int
    limit: int
    offset: int


class DocumentDetailRead(DocumentRead):
    files: list[DocumentFileRead] = []
    pages: list[DocumentPageRead] = []
    chunks: list[DocumentChunkRead] = []
    ocr_jobs: list[OCRJobRead] = []
    audit_logs: list[AuditLogRead] = []


class DocumentMetadataUpdateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=512)
    document_type: str = Field(default="UNKNOWN", max_length=64)
    classification_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    document_number: str | None = Field(default=None, max_length=128)
    document_symbol: str | None = Field(default=None, max_length=128)
    issued_date: date | None = None
    issued_place: str | None = Field(default=None, max_length=255)
    issuing_agency: str | None = Field(default=None, max_length=255)
    excerpt: str | None = None
    recipient: str | None = None
    signer_name: str | None = Field(default=None, max_length=255)
    signer_title: str | None = Field(default=None, max_length=255)
    seals_present: bool | None = None
    attachment_present: bool | None = None
    page_count: int | None = Field(default=None, ge=1)
    business_type: str | None = Field(default=None, max_length=64)


class UploadResponse(BaseModel):
    document: DocumentRead
    ocr_job: OCRJobRead


class MultiFileUploadResponse(BaseModel):
    document: DocumentRead
    files: list[DocumentFileRead]
    ocr_job: OCRJobRead


class SourceFileMutationResponse(BaseModel):
    document: DocumentRead
    files: list[DocumentFileRead]
    ocr_job: OCRJobRead


class ReorderDocumentFilesRequest(BaseModel):
    file_ids: list[str] = Field(min_length=1)


class ReprocessDocumentRequest(BaseModel):
    reason: str | None = None


class ReprocessDocumentResponse(BaseModel):
    document: DocumentRead
    ocr_job: OCRJobRead

File: apps/web/types/document.ts
export interface OCRJob {
  id: string
  document_id: string
  job_type: string
  status: string
  attempts: number
  reason?: string | null
  error_message?: string | null
  created_at: string
  updated_at: string
}

export interface AuditActor {
  id: string
  email: string
  full_name: string
}

export interface AuditLog {
  id: string
  actor_user_id?: string | null
  actor?: AuditActor | null
  action: string
  entity_type: string
  entity_id: string
  metadata: Record<string, unknown>
  created_at: string
}

export interface DocumentItem {
  id: string
  title: string
  original_filename: string
  content_type?: string | null
  document_type: string
  classification_confidence?: number | null
  document_number?: string | null
  document_symbol?: string | null
  issued_date?: string | null
  issued_place?: string | null
  issuing_agency?: string | null
  excerpt?: string | null
  recipient?: string | null
  signer_name?: string | null
  signer_title?: string | null
  seals_present?: boolean | null
  attachment_present?: boolean | null
  page_count?: number | null
  metadata_source?: string | null
  metadata_reviewed_at?: string | null
  business_type?: string | null
  status: string
  created_at: string
  updated_at: string
}

export interface DocumentMetadataInput {
  document_number?: string
  issued_date?: string
  issuing_agency?: string
  business_type?: string
}

export interface DocumentMetadataUpdateInput extends DocumentMetadataInput {
  title: string
  document_type: string
  classification_confidence?: number | null
  document_symbol?: string
  issued_place?: string
  excerpt?: string
  recipient?: string
  signer_name?: string
  signer_title?: string
  seals_present?: boolean | null
  attachment_present?: boolean | null
  page_count?: number | null
}

export interface DocumentPage {
  id: string
  page_number: number
  text: string
  confidence: number
  status: string
}

export interface DocumentFile {
  id: string
  document_id: string
  original_filename: string
  content_type?: string | null
  file_size: number
  file_order: number
  status: string
  created_at: string
  updated_at: string
}

export type SourceFilePreviewMode = 'pdf' | 'image' | 'text'

export interface SourceFilePreview {
  file: DocumentFile
  mode: SourceFilePreviewMode
  object_url: string
  text?: string
}

export interface DocumentChunk {
  id: string
  chunk_index: number
  text: string
  page_from?: number | null
  page_to?: number | null
  section_title?: string | null
  doc_group?: string | null
  chunk_level?: string | null
  section_role?: string | null
  section_path?: string[] | null
  chunk_confidence?: number | null
  requires_review: boolean
  qdrant_point_id?: string | null
}

export interface DocumentDetail extends DocumentItem {
  files: DocumentFile[]
  pages: DocumentPage[]
  chunks: DocumentChunk[]
  ocr_jobs: OCRJob[]
  audit_logs: AuditLog[]
}

export interface UploadResponse {
  document: DocumentItem
  ocr_job: OCRJob
}

export interface MultiFileUploadResponse {
  document: DocumentItem
  files: DocumentFile[]
  ocr_job: OCRJob
}

export interface SourceFileMutationResponse {
  document: DocumentItem
  files: DocumentFile[]
  ocr_job: OCRJob
}

export interface DocumentListFilters {
  q?: string
  status?: string
  document_type?: string
  business_type?: string
  sort_by?: 'created_at' | 'updated_at' | 'issued_date' | 'title' | 'status' | 'document_type' | 'business_type'
  sort_dir?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface ReprocessDocumentResponse {
  document: DocumentItem
  ocr_job: OCRJob
}

export interface ReviewQueueChunk {
  id: string
  document_id: string
  document_title: string
  document_number?: string | null
  issued_date?: string | null
  business_type?: string | null
  chunk_index: number
  text: string
  page_from?: number | null
  page_to?: number | null
  section_title?: string | null
  doc_group?: string | null
  chunk_level?: string | null
  section_role?: string | null
  section_path?: string[] | null
  chunk_confidence?: number | null
  requires_review: boolean
  created_at: string
  updated_at: string
}

export interface ReviewQueueFilters {
  limit?: number
  offset?: number
  section_role?: string
  document_id?: string
  max_confidence?: number | null
}

export interface SearchResult {
  document_id: string
  chunk_id: string
  score: number
  text: string
  title?: string | null
  document_type?: string | null
  document_number?: string | null
  issued_date?: string | null
  business_type?: string | null
  page_from?: number | null
  page_to?: number | null
  doc_group?: string | null
  section_role?: string | null
  section_path: string[]
  requires_review: boolean
}

export interface SemanticSearchFilters {
  limit?: number
  document_type?: string
  business_type?: string
  document_number?: string
  issued_date?: string
  doc_group?: string
  section_role?: string
  requires_review?: boolean | null
}
